### PR TITLE
Add statistics for handle_loaned_message

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Add statistics for handle_loaned_message().   
If loan message is enabled,  some tests in test_subscription_topic_statistics fail.  This issue is found in https://github.com/ros2/rmw_fastrtps/pull/568#issuecomment-1123412147

### CODE REVIEW

File: rclcpp/include/rclcpp/subscription.hpp [hunk 1]
```diff
@@ -363,11 +363,31 @@ class Subscription : public SubscriptionBase
     void * loaned_message,
     const rclcpp::MessageInfo & message_info) override
   {
+    if (matches_any_intra_process_publishers(&message_info.get_rmw_message_info().publisher_gid)) {
+      // In this case, the message will be delivered via intra process and
+      // we should ignore this copy of the message.
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/subscription.hpp:371]: Can you rephrase the questions? I'm not sure I follow what you're asking.

[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/subscription.hpp:371]: > Publisher cannot use LoanedMessage with intra_process_enable.

Yes.  
I only refer to code of handle_message() and not consider whether code is suitable for loan message.  
I will remove above codes.   

For intra-process case, waitable is used for trigger and `dispatch_intra_process()` will be called
https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp#L151-L157

dispatch_intra_process() will call user's callback.   
So I think handle_message() and handle_loaned_message() cannot be called while intra-process is enabled.  

I don't know what case make handle_message() called while  intra-process communication.   
That is,  whether handle_message() needs to include below code.   
```c++
    if (matches_any_intra_process_publishers(&message_info.get_rmw_message_info().publisher_gid)) {
      // In this case, the message will be delivered via intra process and
      // we should ignore this copy of the message.
      return;
    }
```

[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/subscription.hpp]: I'm still not following exactly, I can see the parts you might be concerned with, but I don't follow the specific scenario, or at least I don't see the problem. Maybe we can discuss it in one of the working groups.

[the remark this answers is not in the document]
Comment (wjwwood) [reply rclcpp/include/rclcpp/subscription.hpp]: I actually think the code removed in [565d683](https://github.com/ros2/rclcpp/pull/1927/commits/565d683ca4aa2ddac9f450a38b248939f673a6f7) was useful, but again that might be related to the misunderstanding I have with the above.

AFAICT, it is possible to get the data via loanded message and intra process, so the check to avoid double delivery is valid (I think).

**Review by wjwwood — CHANGES_REQUESTED** [review]
Need to resolve the question about double delivery and the interaction with intra-process.

File: rclcpp/include/rclcpp/subscription.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclcpp/include/rclcpp/subscription.hpp]: > I actually think the code removed in [565d683](https://github.com/ros2/rclcpp/pull/1927/commits/565d683ca4aa2ddac9f450a38b248939f673a6f7) was useful

agree, i believe this code should stay to avoid double delivery. this is what i am saying at the beginning though.

> I'm still not following exactly

sorry, let me add some more comments here. of course we can talk online.

even if publisher does not use `LoanedMessage` in ROS 2, underlying rmw implementation would use shared memory for inter-process communication. in this case, on subscription it could call `handle_loaned_message` if `can_loan_messages` is true. so w/o calling `matches_any_intra_process_publishers` in `handle_loaned_message`, there would be double delivery via rclcpp intra-process and `handle_loaned_message`.  i may be missing something, i am not sure if we can make this happen.

[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/include/rclcpp/subscription.hpp]: I will check what case leads to double delivery.

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya @wjwwood 

This is  a case of double delivery.    
The topic is '/number' for Pub1/Pub2 and Sub1/Sub2.   
Message is Int32 (Pod)  
RMW:  FastDDS  

- Progress A 
   - Pub1 (use_intra_process)
   - Sub 1 (use_intra_process)
- Progress B
   - Pub2

At this time, Sub1 receives messages from Pub1 via dispatch_intra_process(). Sub1 receives messages from Pub2 via handle_loaned_message().   

Now start a new subscriber
- Progress C
   - Sub2 

In order to send message to Sub2, Pub1 will send message via RMW.  
At this time, for the message from Pub1, Sub1 can receive messages via handle_loaned_message() and via dispatch_intra_process(). 
This is double delivery.   

So below code is necessary for handle_loaned_message()
```
    if (matches_any_intra_process_publishers(&message_info.get_rmw_message_info().publisher_gid)) {
      // In this case, the message will be delivered via intra process and
      // we should ignore this copy of the message.
      return;
    }
```

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 

thanks for checking!

> At this time, Sub1 receives messages from Pub1 via dispatch_intra_process().

- Process A
  - Pub1 (use_intra_process, cannot use `LoanedMessage`)
  - Sub1 (use_intra_process)

in this situation, How Sub1 avoid to call `handle_loaned_message`?

my expectation is, Sub1 recognizes `can_loan_messages` is true since data type is PoD. So Sub1 would call `handle_loaned_message` instead of `handle_message`? currently (w/o this PR), `handle_loaned_message` does not call `matches_any_intra_process_publishers`, so i was thinking that there might be double delivery.

i mean there seems to be a double delivery problem in `intra-process` communication.

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya 

> my expectation is, Sub1 recognizes can_loan_messages is true since data type is PoD. So Sub1 would call handle_loaned_message instead of handle_message? currently (w/o this PR), handle_loaned_message does not call matches_any_intra_process_publishers, so i was thinking that there might be double delivery.

Right.   
I forgot to mention ros2/rmw_fastrtps#568 is applied in my test environment. So even if data-sharing is OFF, can_loan_message is true once message is PoD.  
If not apply ros2/rmw_fastrtps#568, handle_message is called instead of handle_loaned_message.

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 yeah, that means we have an unexpected double delivery problem if we enable data sharing for rmw_fastfrpt with current code.

anyway, solution can be applied on this PR.

**Review by fujitatomoya — APPROVED** [review]

**Comment by fujitatomoya** [comment]
we need to backport this to humble, galactic and foxy.

**Comment by fujitatomoya** [comment]
@wjwwood @ivanpauno could you take a look when you have time? I will start the CI.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16718/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11298/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17112/)

**Review by wjwwood — APPROVED** [review]
lgtm

**Comment by fujitatomoya** [comment]
@mergify backport humble galactic foxy
